File: src-tauri/src/model/activation.rs
use serde::{Deserialize, Serialize};

use super::day_time::{DaysOfMonth, DaysOfWeek};

#[derive(Debug, Serialize, Deserialize, Clone)]
pub enum ActivationOperator {
	AND,
	OR
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(tag = "type")]
pub enum ActivationQuery {
	OnDaysOfWeek(DaysOfWeek),
	OnDaysOfMonth(DaysOfMonth)
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct ActivationModifier {
	pub query: ActivationQuery,
	pub operator: ActivationOperator
}

impl ActivationModifier {
	pub fn new(
		query: ActivationQuery,
		operator: ActivationOperator
	) -> Self {
		return Self { query, operator };
	}
}

File: src/backend/activation.ts
import { DaysOfMonth, DaysOfWeek } from "./dayTime";

export type ActivationOperator = "AND" | "OR";

interface OnDaysOfWeekActivationQuery extends DaysOfWeek {
    tag: "OnDaysOfWeek"
}

interface OnDaysOfMonthActivationQuery extends DaysOfMonth {
    tag: "OnDaysOfMonth"
}

export type ActivationQuery = | OnDaysOfWeekActivationQuery
                              | OnDaysOfMonthActivationQuery;

export interface ActivationModifier {
    query: ActivationQuery,
    operator: ActivationOperator
}
